impeller/docs/faq.md

```diff
@@ -100,3 +100,133 @@
+    inherent in the way Flutter held and used our original renderer. But, in
+    truth, the mandate to work on Impeller would never have materialized if it
+    weren't for the issue of shader compilation jank. In-spite of the massive
+    sunk cost of attempting all the workarounds shader complication jank that
+    prevented us from delivering our product requirement to deliver consistent
+    60fps even at first-run.
+  * A very early look at how Impeller avoids the problems in the previous
+    renderer are [discussed in this
+    tech-talk](https://www.youtube.com/watch?v=gKrYWC_SDxQ). Our current
+    renderer is way more advanced today but that talk describes the very early
+    thinking, motivations, and decisions.
+  * We turned on Impeller by default for iOS in early 2023, and it has been in
+    production in the majority of iOS apps since then. It's currently, as of
+    3.22 an opt-in option for Flutter on Android.
+  * Since we enabled Impeller on by default for iOS, we have seen the vast
+    majority of open issues around shader compilation causing jank to have been
+    fixed. We know there is more work to do, but we're encouraged by the results
+    we've been hearing from our customers.
+  * Today, Impeller outperforms the old renderer not only in worst-frame timed
+    benchmarks, but is also faster on average.
+* What makes Impeller different, compared to Skia, for Flutter?
+  * We want to start by saying Skia is amazing. We'd also never say Impeller is
+    better than Skia. It just works differently. The original design of Impeller
+    was based on collaboration with the Skia team and was greatly influenced by
+    their feedback. We also keep collaborating on an ongoing basis. Ideas such
+    as stencil-then-cover that Impeller now uses originated from Skia. Flutter
+    also continues to use Skia for text layout and its image codecs and has no
+    plans to migrate away from using those sub-components. We'd wholeheartedly
+    recommend Skia for most rendering needs.
+  * All of Impellers shaders are [manually authored and
+    compiled](https://github.com/flutter/engine/tree/0a8de3dd3285c0b64de47630a8218ae38b8e04e1/impeller#the-offline-shader-compilation-pipeline)
+    during build time and packaged with the Flutter engine. There is no runtime
+    shader generation, reflection, or compilation. Skia can and does generate
+    and compile shaders at runtime.
+  * By necessity, Impeller has a bounded set of shaders (< 50) that are known
+    ahead of time. Due to the way rendering intent is parameterized, Impeller
+    also needs way fewer shaders than Skia. All the graphics pipelines needed by
+    Impeller are ready well before the Flutter application's Dart isolate is
+    launched. Skia can generate and compile shaders during frame workloads
+    leading to worse worst-frame times.
+  * Because the processing of shaders in Impeller happens at build time when
```

Comment:
As a minor nit, we are still compiling shaders at runtime, and may generate new PSOs on demand. But we don't have to go through a source -> source translation or reflection step, and on iOS we can use metal achive files which load much faster.

The GPU is still jitting the shader though

Reply:
Sounds like a good thing to mention. Is there a general notion we can give of the relative cost between PSO generation and the source->source translation? Like is it 10x cheaper?

Reply:
100x maybe?

Reply:
on iOS in particular, it seems to be closer to "free"

Reply:
(The PSO generation)

Reply:
And important to point out this isn't an alternative, in the old route we go source -> source -> metal compiler (source) -> PSO. now we just go metal compiler (archive) -> PSO.

Reply:
I went back and forth quite a bit on this and adding backend specific nuance made things a whole lot more complicated. On Metal, we can use either AIR files or the new Metal Binary Archives which really move everything to be at build time. But have to resort to flinging strings around with OpenGL ES. So what happens before GPU microcode is generated is quite quite messy to explain. This seemed like a fine enough bit of context to tell the story. Should I add another FAQ entry perhaps?

@zanderso From go/impeller-prototype, I found this screenshot of a trace I took where the cost of the shader library construction is separate from the PSO creation (the small pink trace at the end). I think in this version of Skia, this included source-to-source translation as well as the shader library construction we do [here in Impeller](https://github.com/flutter/engine/blob/15b1deab952ee114c06ddccfa005e2df82c0d35d/impeller/tools/metal_library.py#L72). Even that last bit at the end, Impeller does before first frame (well variants are another matter I suppose).

<img width="1774" alt="Screenshot 2024-08-09 at 10 56 02 AM" src="https://github.com/user-attachments/assets/0add12d3-94b9-46bb-bbf5-d5a478fee321">

Reply:
Oh, nevermind. In this case, there is an SKSL cache hit. Its even slower when there is no cached SKSL. This used to happen when the phone was rebooted and Metals caches weren't warmed up from previous runs. I was making that case that perfect SKSL caches were necessary but not sufficient for jank free operation because metalc literally had to do LLVM things and that was more than sufficient to blow our frame budgets.

Such a mess and I've blocked away so much of this context in my head.

Reply:
We, of course, never found this janky behavior on our benchmarks because we didn't reboot phones between runs to clear the internal metal caches.